marshal.go

```diff
@@ -5,42 +5,66 @@ package fargo
 import (
 	"encoding/json"
 	"encoding/xml"
+	"fmt"
 	"io"
 	"strconv"
 )
 
-// Temporary structs used for GetAppsResponse unmarshalling
-type getAppsResponseArray GetAppsResponse
-type getAppsResponseSingle struct {
-	Application   *Application `json:"application"`
-	AppsHashcode  string       `json:"apps__hashcode"`
-	VersionsDelta int          `json:"versions__delta"`
+func intFromJSONNumberOrString(jv interface{}, description string) (int, error) {
+	switch v := jv.(type) {
+	case float64:
+		return int(v), nil
+	case string:
+		n, err := strconv.Atoi(v)
+		if err != nil {
+			return 0, err
+		}
+		return n, nil
+	default:
+		return 0, fmt.Errorf("unexpected %s: %[2]v (type %[2]T)", description, jv)
+	}
 }
 
 // UnmarshalJSON is a custom JSON unmarshaler for GetAppsResponse to deal with
 // sometimes non-wrapped Application arrays when there is only a single Application item.
 func (r *GetAppsResponse) UnmarshalJSON(b []byte) error {
 	marshalLog.Debugf("GetAppsResponse.UnmarshalJSON b:%s\n", string(b))
-	var err error
+	resolveDelta := func(d interface{}) (int, error) {
+		return intFromJSONNumberOrString(d, "versions delta")
+	}
 
 	// Normal array case
-	var ra getAppsResponseArray
-	if err = json.Unmarshal(b, &ra); err == nil {
-		marshalLog.Debug("GetAppsResponse.UnmarshalJSON ra:%+v\n", ra)
-		*r = GetAppsResponse(ra)
-		return nil
+	type getAppsResponse GetAppsResponse
+	auxArray := struct {
+		*getAppsResponse
+		VersionsDelta interface{} `json:"versions__delta"`
+	}{
+		getAppsResponse: (*getAppsResponse)(r),
 	}
+	var err error
+	if err = json.Unmarshal(b, &auxArray); err == nil {
+		marshalLog.Debugf("GetAppsResponse.UnmarshalJSON array:%+v\n", auxArray)
+		r.VersionsDelta, err = resolveDelta(auxArray.VersionsDelta)
```

Comment:
Note line 42, where we populate the "auxArray" variable's "getAppsResponse" [(anonymous|embedded) field](https://golang.org/ref/spec#AnonymousField) with variable "r." When we supply "auxArray' to `json.Unmarshal` at line 45, it will write directly to a "r" (dereferencing the pointer as necessary), which is analogous to the previous assignment you recall. It will write to "r" _and_ also write into the auxiliary struct's "VersionsDelta" field, which we then convert and assign back into "r"'s "VersionsDelta" field. The outer struct's field (of type `interface{}`) shadows [the field with the same name in the embedded "getAppsResponse" field](https://github.com/seh/fargo/blob/f1f7c88ad0e3cca61c8a774c5f3c58f3fae53498/struct.go#L40) (of type `int`).

Here the variable name "auxArray" means "auxiliary structure for the "array of applications" case, versus the "single application" case handled at line 52. The variable name "aux" comes up frequently with this encoding or decoding technique in Go, where one introduces some separate, temporary, and only locally significant struct to handle the finicky details that Go's struct tags can't accommodate.